File: TourPlanner.BL/TourService.cs
namespace TourPlanner.BL;

using TourPlanner.DAL;
using TourPlanner.Models;

public class TourService : ITourService
{
    private readonly ITourRepository _repository;

    public TourService(ITourRepository repository)
    {
        _repository = repository;
    }

    public IEnumerable<Tour> GetAll() => _repository.GetAll();
    public IEnumerable<Tour> GetAllByUserId(Guid userId) => _repository.GetAllByUserId(userId);
    public Tour? GetById(Guid id) => _repository.GetById(id);
    public Tour Add(Tour tour) => _repository.Add(tour);
    public Tour? Update(Tour tour) => _repository.Update(tour);
    public bool Delete(Guid id) => _repository.Delete(id);
}

File: tourplanner-frontend/src/app/services/tour.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Tour } from '../models/tour.model';

@Injectable({ providedIn: 'root' })
export class TourService {
  private readonly apiUrl = 'http://localhost:5210/api/tours';

  constructor(private http: HttpClient) {}

  getAll(): Observable<Tour[]> {
    return this.http.get<Tour[]>(this.apiUrl);
  }

  getById(id: string): Observable<Tour> {
    return this.http.get<Tour>(`${this.apiUrl}/${id}`);
  }

  create(tour: Partial<Tour>): Observable<Tour> {
    return this.http.post<Tour>(this.apiUrl, tour);
  }

  update(id: string, tour: Partial<Tour>): Observable<Tour> {
    return this.http.put<Tour>(`${this.apiUrl}/${id}`, tour);
  }

  delete(id: string): Observable<void> {
    return this.http.delete<void>(`${this.apiUrl}/${id}`);
  }
}
